libarchive/test/test_write_format_zip_compression_zstd.c

```diff
@@ -0,0 +1,424 @@
+/*-SPDX-License-Identifier: BSD-2-Clause
+ * Copyright (c) 2024 ARJANEN Loïc Jean David
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "test.h"
+#ifdef HAVE_ZSTD_H
+#include <zstd.h>
+
+static unsigned long
+bitcrc32(unsigned long c, const void *_p, size_t s)
+{
+	/* This is a drop-in replacement for crc32() from zlib, given that
+	 * libzstd doesn't use CRC32 in the first place, let alone has a
+	 * function for it. Libarchive should be able to correctly generate
+	 * ZSTD-compressed zip archives (including correct CRCs) even when
+	 * zlib is unavailable, and this function helps us verify that. Yes,
+	 * this is very, very slow and unsuitable for production use, but
+	 * it's very, very obviously correct, compact, and works well for
+	 * this particular usage. Libarchive internally uses a much more
+	 * efficient implementation when zlib is unavailable */
+	const unsigned char *p = _p;
+	int bitctr;
+
+	if (p == NULL)
+		return (0);
+
+	for (; s > 0; --s) {
+		c ^= *p++;
+		for (bitctr = 8; bitctr > 0; --bitctr) {
+			if (c & 1) c = (c >> 1);
+			else	   c = (c >> 1) ^ 0xedb88320;
+			c ^= 0x80000000;
+		}
+	}
+	return (c);
+}
```

Comment:
Could we use the crc32 function in libarchive/archive_crc32.h instead? Other test files seem to use that.

Reply:
I followed the example of the Store compression test in using a very straightforward CRC-32 implementation for the tests (except the LZMA/XZ ones, where we are sure to have liblzma's implementation), but I'm not wed to it.

Reply:
`archive_crc32.h` is an internal libarchive implementation detail.  Tests should try not to rely on libarchive internals.

Reply:
I would normally agree on not depending on implementation details, but in this case I would prefer to avoid code duplication or going with a significantly suboptimal version.

Reply:
So maybe moving this function to test_utils so that any test needing a CRC32 and unable to guarantee having zlib/liblzma/liblzo at hand could be sure to have one?

But that would mean changing the Store test too. And, for good measure, things like the RAR5 format tests which don't bother with creating a straightforward CRC32 function and just pick the one in archive_crc32.h.

Reply:
Moving this function to test_utils and migrating other tests to use it would make a nice follow-up PR.

Reply:
I think it's important that the tests use a different implementation: Someday, someone is going to decide to improve the implementation in archive_crc32.h and if the tests are also using that same code, then errors in it are going to not be obvious.

That said, once we've consolidated the tests to all use a single implementation in test_utils, we could replace that implementation with something more efficient in order to speed up testing.  As long as the tests and libarchive internals are using different implementations, we can replace/upgrade either one and testing the two against each other will help ensure that both are correct.